### PR TITLE
Default Deep Link Handler

### PR DESCRIPTION
**Pull request by sshropshire**
### Summary of changes

 - Add `BraintreeClient#useDefaultDeepLinkHandler()` boolean flag to opt in to using the SDKs default deep link handler
 - Create `BraintreeDeepLinkActivity` to capture deep link results when default deep link behavior is enabled

 ### Checklist

 - [x] Added a changelog entry

- @jaxdesmarais
- @sshropshire

### CODE REVIEW

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClient.java [hunk 1]
```diff
@@ -331,4 +360,21 @@ Context getApplicationContext() {
...
+    /**
+     * Set this property to true to enable default deep link handling for browser switched flows.
+     * For web flows, this means launching the browser in a task separate from the calling activity.
+     *
+     * NOTE: When default deep linking is enabled, all custom url schemes set in {@link BraintreeClient}
+     * constructors will be ignored.
+     *
+     * @param useDefaultDeepLinkHandler set to true to enable the default deep link handler. This value is false by default.
+     */
+    public void useDefaultDeepLinkHandler(boolean useDefaultDeepLinkHandler) {
```
Comment (scannillo) [inline BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClient.java:377]: Calling the behavior from V3 the `default` behavior is confusing to me. Especially since merchants will have to opt-in to enable the "default" behavior.

Would it be more clear to call this something like `launchBrowsersAsNewTask(true)`?

Comment (sshropshire) [reply BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClient.java:377]: Yeah I'm up for renaming this. "Default Deep Link Handler" is the sentiment, but it is confusing – if this is set to `false` there is actual _default_ SDK behavior that already exists.

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClient.java [hunk 2]
```diff
@@ -331,4 +360,21 @@ Context getApplicationContext() {
...
+
+    /**
+     * Set this property to true to enable default deep link handling for browser switched flows.
+     * For web flows, this means launching the browser in a task separate from the calling activity.
+     *
+     * NOTE: When default deep linking is enabled, all custom url schemes set in {@link BraintreeClient}
+     * constructors will be ignored.
+     *
+     * @param launchesBrowserSwitchAsNewTask set to true to enable the default deep link handler. This value is false by default.
+     */
```
Comment (scannillo) [inline BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClient.java:376]: Nitpick - I think we can reword this to remove `default` references, given the new property name.

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClientParams.java [hunk 3]
```diff
@@ -18,6 +18,7 @@ class BraintreeClientParams {
     private BrowserSwitchClient browserSwitchClient;
     private ManifestValidator manifestValidator;
     private UUIDHelper uuidHelper;
+    private String defaultReturnURLScheme;
```
Comment (scannillo) [inline BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClientParams.java:21]: Nitpick - internal naming still is using `default`. Do we want to change this to closer mirror the new public property naming? Like `newTaskReturlURLScheme`?

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package com.braintreepayments.api;
+
+import android.os.Bundle;
+
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AppCompatActivity;
+
+/**
+ * This Activity is used by the SDK Activity to enable default deep link handling and should not
```
Comment (scannillo) [inline BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java:9]: Same comment here re: `default` wording

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * be used directly.
+ */
+public class BraintreeDeepLinkActivity extends AppCompatActivity {
```
Comment (scannillo) [inline BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java:12]: Is the reason this is public so merchants can perform any clean-up on this activity if need be?

File: LocalPayment/src/test/java/com/braintreepayments/api/LocalPaymentLifecycleObserverUnitTest.java [hunk 6]
```diff
@@ -70,6 +110,23 @@ public void onResume_whenPendingBrowserSwitchResultExists_andRequestCodeNotPayPa
 
         sut.onStateChanged(activity, Lifecycle.Event.ON_RESUME);
 
-        verify(localPaymentClient, never()).onBrowserSwitchResult(any(FragmentActivity.class));
+        verify(localPaymentClient, never()).onBrowserSwitchResult(any(FragmentActivity.class), any(BrowserSwitchResult.class));
+    }
+
+    @Test
+    public void onResume_whenCachedBrowserSwitchResultExists_andRequestCodeNotPayPal_doesNothing() {
```
Comment (scannillo) [inline LocalPayment/src/test/java/com/braintreepayments/api/LocalPaymentLifecycleObserverUnitTest.java:117]: 👏 For making these tests really descriptive and easy to read.

Comment (scannillo) [reply LocalPayment/src/test/java/com/braintreepayments/api/LocalPaymentLifecycleObserverUnitTest.java:117]: Should this name be `_andRequestCodeNotLocalPayment_`?

File: PayPal/src/test/java/com/braintreepayments/api/PayPalClientUnitTest.java [hunk 7]
```diff
@@ -240,6 +241,34 @@ public void tokenizePayPalAccount_startsBrowser() throws JSONException, BrowserS
         assertEquals("paypal-browser", metadata.get("source"));
     }
 
+    @Test
+    public void tokenizePayPalAccount_whenDefaultDeepLinkHandlerEnabled_startsBrowserAsSingleTask() throws JSONException, BrowserSwitchException {
```
Comment (scannillo) [inline PayPal/src/test/java/com/braintreepayments/api/PayPalClientUnitTest.java:245]: Same comment throughout re: internal naming

```suggestion
    public void tokenizePayPalAccount_whenLaunchesBrowserSwitchAsNewTaskEnabled_startsBrowserAsSingleTask() throws JSONException, BrowserSwitchException {
```

Comment (scannillo) [reply PayPal/src/test/java/com/braintreepayments/api/PayPalClientUnitTest.java:245]: We could also 🏈 these internal changes to a separate PR after the release too

File: LocalPayment/src/test/java/com/braintreepayments/api/LocalPaymentLifecycleObserverUnitTest.java (hunk 6 above)
Comment (scannillo) [inline LocalPayment/src/test/java/com/braintreepayments/api/LocalPaymentLifecycleObserverUnitTest.java:117]:
```suggestion
    public void onResume_whenCachedBrowserSwitchResultExists_andRequestCodeNotLocalPayment_doesNothing() {
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,9 @@
 * PayPalNativeCheckoutClient
   * Bumping native-checkout version to 0.8.1 
   * Adding in Native checkout support for one time password
+* BraintreeCore
+  * Add `BraintreeClient#launchesBrowserSwitchAsNewTask()` boolean flag to opt in to using the SDKs default deep link handler
```
Comment (scannillo) [inline CHANGELOG.md:9]: Same comment re: removing `default` wording.

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClientParams.java (hunk 3 above)
Comment (sshropshire) [reply BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeClientParams.java:21]: Yeah this makes sense. We may go with `braintreeDeepLinkReturnUrlScheme`. Also added a `NEXT MAJOR VERSION` note to formalize capitalization of `URL` etc. since at the moment other properties have a `Url` convention that isn't consistent.

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java (hunk 5 above)
Comment (sshropshire) [reply BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeDeepLinkActivity.java:12]: This is a good callout we did a quick test. It does need to be public, I think because it's a deep link target the Android system needs access to the class definition in order to instantiate it.

**Comment by sshropshire** [comment]
@scannillo all good! Thanks for the feedback we were able to make those changes. We'll have an opportunity to clean up the naming more in `v5`. The plan is to make this feature actual default behavior in the next major version.

**Review by demerino — APPROVED** [review]
Looks good
